File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-3.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> dp(cost.size()+1, -1);
        return min(dfs(cost, dp, 0), dfs(cost,dp, 1));
    }
    int dfs(vector<int>& cost, vector<int> & dp, int i){
        if(i >= cost.size()){
            return 0;
        }
        if(dp[i]!= -1){
            return dp[i];
        }
        dp[i] = cost[i] + min(dfs(cost, dp, i+1), dfs(cost, dp, i+2));
        return dp[i];
    }
};

File: Data Structures & Algorithms/course-schedule/submission-3.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        pre = defaultdict(list)
        for p in prerequisites:
            pre[p[1]].append(p[0])
        visit = set()
        def dfs(i):
            if pre[i] == []:
                return True
            if i in visit:
                return False
            visit.add(i)
            for p in pre[i]:
                if not dfs(p): return False
            visit.remove(i)
            pre[i]=[]
            return True
        for i in range(numCourses):
            if not dfs(i):
                return False
        return True
            
